noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays.nr

```diff
@@ -143,323 +141,6 @@ pub fn check_permutation<T, let N: u32>(
-    run_lengths: [u32; N]
-) where T: Positioned + Ordered + Empty + Eq {
-    /*
-    The original_array here needs to be sorted based on the `position` field of the container,
-    *and* a secondary sort based on the `counter` field of the container. 
-
-    For example, the storage slot in the case of public data update requests.
-    The run_lengths array should contain the length of each run of the original_array.
-    The deduped_array should contain the deduplicated array.
-
-    For example, if the original array is writing `(position,value,counter)`s:
-    [ (1,1,1), (1,2,4), (2,3,3), (3,4,2), (3,5,5), (3,6,6), (4,7,8), (4,8,9), (5,9,7), (0,0,0), ... padding with zeros ]
-    then run_lengths array is:
-    [
-        2, // run of 1s
-        1, // run of 2
-        3, // run of 3s
-        2, // run of 4s
-        1, // run of 5
-        0,
-        0,
-        ... padding with zeros
-    ]
-    
-    then the deduped_array should be:
-    [ (1,2,4), (2,3,3), (3,6,6), (4,8,9), (5,9,7), (0,0,0), ... padding with zeros ]
-    */
-
-    let deduped_len = validate_array(deduped_array); // This also makes sure that the array is padded with empty items.  
-    let run_lengths_len = array_length(run_lengths); // Don't have to be a "validated" array because non zero padded values don't hurt.  
-    assert_eq(deduped_len, run_lengths_len, "Deduped array length does not match number of run lengths");
-
-    let mut seen_empty = false;
-    // container at the start of the current run
-    let mut start_run_index = 0;
-    // the index we are collapsing into
-    let mut deduped_index = 0;
-    // the length of the current run we are collapsing
-    let mut run_counter = run_lengths[deduped_index];
-    for i in 0..N {
-        let current_container = original_array[i];
-        if is_empty(current_container) {
-            seen_empty = true;
-        } else {
-            assert(!seen_empty, "Empty values must be padded to the right");
-            assert(run_counter > 0, "Invalid run length");
-            assert(
-                current_container.position().eq(original_array[start_run_index].position()), "The position of the current container must match the start of the run"
-            );
-            run_counter -= 1;
-            if run_counter == 0 {
-                assert(
-                    deduped_array[deduped_index].eq(current_container), "The container we are collapsing into must match the current container"
-                );
-                start_run_index = i + 1;
-                deduped_index += 1;
-                run_counter = run_lengths[deduped_index];
-            } else {
-                // we're in a run, so this container must have a lower counter.
-                // note we don't check for overflow here, as the run_lengths array must be correct.
-                assert(
-                    current_container.counter() <= original_array[i + 1].counter(), "Containers in a run must be sorted by counter"
```

Comment:
The old code only checked that items in each run meet the requirement - same position with incrementing counters. But we can split items of the same position into two runs and it would pass.